### PR TITLE
BHoM: clarification of concept of Hash vs Diff; simplification of Query.Hash()

### PR DESCRIPTION
**Pull request by alelom**
### Issues addressed by this PR

Closes https://github.com/BHoM/BHoM/issues/1083

This brings a better distinction between Diffing and Hash and uniforms the required parameters to compute both.

For example, previously the Diffing required a Tolerance (expressed as in e.g. `1E-6`) while `Query.Hash()` required a "fractionalDigits" input (expressed as in e.g. `6`).

This PR together with https://github.com/BHoM/BHoM_Engine/pull/2105 brings consistency and makes things more reliable.

### CODE REVIEW

**Comment by alelom** [comment]
@BHoMBot CheckCore

**Comment by alelom** [comment]
@BHoMBot check core

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

File: BHoM/ComparisonConfig.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        public List<string> NamespaceExceptions { get; set; } = new List<string>(); //e.g. `BH.oM.Structure`
+
+        [Description("Any corresponding type is ignored. E.g. `typeof(Guid)`.")]
+        public List<Type> TypeExceptions { get; set; } = new List<Type>(); //e.g. `typeof(Guid)`
+
+        [Description("Keys of the BHoMObjects' CustomData dictionary that should be ignored.\nBy default it includes `RenderMesh`.")]
+        public virtual List<string> CustomdataKeysExceptions { get; set; } = new List<string>() { "RenderMesh" };
+
+        [Description("If any property is nested into the object over that level, it is ignored. Defaults to 100.")]
+        public int MaxNesting = 100;
```
Comment (FraserGreenroyd) [inline BHoM/ComparisonConfig.cs:56]: Should be virtual, and have get/set accessors

File: BHoM/ComparisonConfig.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+        [Description("Numeric tolerance for property values, applied to all numerical properties. Applies rounding for numbers smaller than this. Defaults to 1E-12.")]
+        public double NumericTolerance = 1E-12;
```
Comment (FraserGreenroyd) [inline BHoM/ComparisonConfig.cs:59]: Should be virtual, and have get/set accessors

File: BHoM/ComparisonConfig.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+        [Description("Number of fractional digits retained for individual property. If a property name matches a key in the dictionary, applies a rounding to the corresponding number of digits."
+            + "\nSupports * wildcard in the property name matching. E.g. `{ { StartNode.Point.*, 2 } }`.")]
+        public Dictionary<string, int> FractionalDigitsPerProperty = null; // e.g. { { StartNode.Point.X, 2 } } � can use * wildcard here.
```
Comment (FraserGreenroyd) [inline BHoM/ComparisonConfig.cs:63]: Should be virtual, and have get/set accessors

File: Diffing_oM/DiffConfig/DiffingConfig.cs [hunk 4]
```diff
@@ -31,45 +31,31 @@
...
             "\nDefaults to 1000.")]
         public virtual int MaxPropertyDifferences { get; set; } = 1000;
 
-        [Description("If enabled, the Diff stores also the objects that did not change (`Unchanged` property).")]
-        public virtual bool StoreUnchangedObjects { get; set; } = true;
+        [Description("If enabled, the Diff includes also the objects that did not change (`Unchanged`).")]
+        public virtual bool IncludeUnchangedObjects { get; set; } = true;
+
+        [Description("Type of PersistentId that should be used to perform the Diffing.")]
+        public virtual Type PersistentIdType { get; set; }
```
Comment (FraserGreenroyd) [inline Diffing_oM/DiffConfig/DiffingConfig.cs:55]:
```suggestion
        [Description("Type of PersistentID that should be used to perform the Diffing.")]
        public virtual Type PersistentIDType { get; set; }
```

File: Diffing_oM/DiffConfig/DiffingConfig.cs [hunk 5]
```diff
@@ -31,45 +31,31 @@
...
+
+        [Description("Key of the CustomData dictionary where to look for an Id to use for the Diffing.")]
```
Comment (FraserGreenroyd) [inline Diffing_oM/DiffConfig/DiffingConfig.cs:57]:
```suggestion
        [Description("Key of the CustomData dictionary where to look for an ID to use for the Diffing.")]
```

The sentence as a whole does not make a lot of sense to me? Maybe it's cause it's late so I might not be comprehending it correctly but it doesn't read right to me?

File: Diffing_oM/DiffingType.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        [Description("Revision: for the Stream-revision workflow. The Diffing considers different Revisions of a Stream (a set of object that changed in time).")]
+        Revision,
+        [Description("PersistentId: uses the PersistentId stored in the Fragments of the BHoMObjects, if present. If more than one PersistentId is found on the objects, a PersistentId type must be specified in the DiffConfig.")]
+        PersistentId,
+        [Description("OneByOne: assuming the input collections are of the same length, this compares objects one by one.")]
+        OneByOne,
+        [Description("Hash: compares by using the object hashes. If there is already a HashFragment stored in the BHoMObjects, by default that is used. Otherwise, hash is computed for every object, and then by default it gets stored in the BHoMObjects. Default behaviours can be changed in the DiffConfig.")]
+        Hash,
+        [Description("CustomData: uses an Id stored under the BHoMObjects CustomData.")]
+        CustomDataId
```
Comment (FraserGreenroyd) [inline Diffing_oM/DiffingType.cs:47]:
```suggestion
        CustomDataID
```

File: Diffing_oM/DiffingType.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using BH.oM.Base;
+using System.ComponentModel;
+
+namespace BH.oM.Diffing
+{
+    [Description("What type of Diffing should be computed.")]
+    public enum DiffingType
```
Comment (FraserGreenroyd) [inline Diffing_oM/DiffingType.cs:34]: Enum should be alphabetical - and should have `Undefined` as a starting option unless a very strong reason against doing so can be provided.

File: Diffing_oM/DiffingType.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        Automatic,
+        [Description("Revision: for the Stream-revision workflow. The Diffing considers different Revisions of a Stream (a set of object that changed in time).")]
+        Revision,
+        [Description("PersistentId: uses the PersistentId stored in the Fragments of the BHoMObjects, if present. If more than one PersistentId is found on the objects, a PersistentId type must be specified in the DiffConfig.")]
+        PersistentId,
+        [Description("OneByOne: assuming the input collections are of the same length, this compares objects one by one.")]
+        OneByOne,
+        [Description("Hash: compares by using the object hashes. If there is already a HashFragment stored in the BHoMObjects, by default that is used. Otherwise, hash is computed for every object, and then by default it gets stored in the BHoMObjects. Default behaviours can be changed in the DiffConfig.")]
+        Hash,
+        [Description("CustomData: uses an Id stored under the BHoMObjects CustomData.")]
```
Comment (FraserGreenroyd) [inline Diffing_oM/DiffingType.cs:46]:
```suggestion
        [Description("CustomData: uses an ID stored under the BHoMObjects CustomData.")]
```

File: Diffing_oM/DiffingType.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+{
+    [Description("What type of Diffing should be computed.")]
+    public enum DiffingType
+    {
+        [Description("Automatic: tries, in sequence, diffing with PersistentId, OneByOne, Hash.")]
+        Automatic,
+        [Description("Revision: for the Stream-revision workflow. The Diffing considers different Revisions of a Stream (a set of object that changed in time).")]
+        Revision,
+        [Description("PersistentId: uses the PersistentId stored in the Fragments of the BHoMObjects, if present. If more than one PersistentId is found on the objects, a PersistentId type must be specified in the DiffConfig.")]
+        PersistentId,
```
Comment (FraserGreenroyd) [inline Diffing_oM/DiffingType.cs:41]:
```suggestion
        [Description("PersistentId: uses the PersistentID stored in the Fragments of the BHoMObjects, if present. If more than one PersistentId is found on the objects, a PersistentID type must be specified in the DiffConfig.")]
        PersistentID,
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
I have reviewed all of BHoMBot's suggestions and agree with all of them, and have added some more that BHoMBot doesn't yet pick up on.

**Comment by al-fisher** [comment]
> On the above Id vs ID

Must admit I do prefer the form ID - for similar reasons stated in @FraserGreenroyd’s linked discussion.

File: Diffing_oM/Diff.cs [hunk 10]
```diff
@@ -62,12 +62,12 @@ public class Diff : IObject, IImmutable
...
         [Input("modifiedObjects", "Objects existing in both sets that have some differences in their properties.")]
         [Input("modifiedPropsPerObject", "Dictionary holding the differences in properties of the 'modified' objects. See the corresponding property description for more info.")]
-        public Diff(IEnumerable<object> addedObjects, IEnumerable<object> removedObjects, IEnumerable<object> modifiedObjects, DiffConfig diffConfig, Dictionary<string, Dictionary<string, Tuple<object, object>>> modifiedPropsPerObject = null, IEnumerable<object> unchangedObjects = null)
+        public Diff(IEnumerable<object> addedObjects, IEnumerable<object> removedObjects, IEnumerable<object> modifiedObjects, DiffingConfig diffConfig, Dictionary<string, Dictionary<string, Tuple<object, object>>> modifiedPropsPerObject = null, IEnumerable<object> unchangedObjects = null)
         {
             AddedObjects = addedObjects;
             RemovedObjects = removedObjects;
             ModifiedObjects = modifiedObjects;
-            DiffConfig = diffConfig;
+            DiffingConfig = diffConfig;
```
Comment (IsakNaslundBh) [inline Diffing_oM/Diff.cs:70]: Input parameter must have the same of the property for serialisation to work.

File: Diffing_oM/DiffingType.cs (hunk 7 above)
Comment (IsakNaslundBh) [reply Diffing_oM/DiffingType.cs:34]: Disagree:P

Comment (adecler) [reply Diffing_oM/DiffingType.cs:34]: A bit worried to get into that conversation but I would say `Automatic` is a pretty good default value for an enum. 

On an anecdotical level, I personally prefer to sort my enums alphabetically (when I remember to do so at least) but would not impose it on others.

Comment (al-fisher) [reply Diffing_oM/DiffingType.cs:34]: `Automatic` as the leading enum and default, makes sense to me

Comment (al-fisher) [reply Diffing_oM/DiffingType.cs:34]: Agree with Isak - think alphabetical and forcing undefined - is too restrictive and does not make sense in this case.

File: Diffing_oM/DiffConfig/DiffingConfig.cs (hunk 5 above)
Comment (alelom) [reply Diffing_oM/DiffConfig/DiffingConfig.cs:57]: Sentence is ok to me, it means what it says. See https://github.com/BHoM/BHoM_Engine/issues/2073#issuecomment-719507129 to the referred use case.

Comment (FraserGreenroyd) [reply Diffing_oM/DiffConfig/DiffingConfig.cs:57]: Fair enough, in that case just the `Id` => `ID` needs changing :smile:

File: Diffing_oM/Diff.cs (hunk 10 above)
Comment (alelom) [reply Diffing_oM/Diff.cs:75]: Thanks @IsakNaslundBh

**Comment by alelom** [comment]
I would leave the Id versus ID renaming to another time, as it could be a significant exercise which at this point there is little time to implement.

**Comment by al-fisher** [comment]
> I would leave the Id versus ID renaming to another time, as it could be a significant exercise which at this point there is little time to implement. I would prefer keeping consistency within the Diffing at the moment.
> 
> Raised issue for this: #1107

Perfect - agreed. Thanks for raising issue

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by alelom** [comment]
@BHoMBot check compliance

**Comment by alelom** [comment]
@BHoMBot check compliance

**Comment by alelom** [comment]
@BHoMBot check compliance

**Comment by alelom** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check installer

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check installer

**Review by FraserGreenroyd — APPROVED** [review]
I'm now happy, approving in line with other PRs, would be happy for this to be merged in its current state following discussions with @alelom :smile:

**Pull request merged**